### ISSUE TITLE
Guidance for implementing Dynamic Routing

### ISSUE THREAD

**Issue by darylrobbins** [spring-cloud/spring-cloud-consul#271]
**@spencergibb -- Big fan of all the great work that you've been doing on Spring Cloud and hoping you can provide a bit of help on this question. For our application, we need dynamic routing for blue/green deployments, which is not currently supported. I am looking at implementing (and contributing) this functionality. I would appreciate any guidance on what you think the design of dynamic routing would look like.**

From other issues in the project, I am aware that dynamic routing is not currently supported with Spring Cloud Netflix/Consul. This is a feature that is very important to us from a blue/green deployment perspective, which is why I am looking at how we could potentially implement this functionality (and contribute back to the project). Given your extensive experience with this complicated infrastructure, I was hoping you could provide me with some guidance on how you would envision this functionality being implemented. I am sure you must have a rough design in your mind of how this would look and I would hate to re-invent the wheel on this.

We essentially want to run two versions of the application simultaneously behind Zuul. General traffic would be routed to the old version, while test traffic could be directed at the new version (using an HTTP header, for example). Once testing of the new version was complete, we'd tell the routers to direct all traffic to the new version.

I have spent the greater part of a day tracing through the code both in my editor and my debugger. There is definitely an elegance to how all the components integrate together into a cohesive whole, but the loosely-coupled nature of Spring can make certain elements very hard to follow all the magic that is going on (especially when you've got nearly 5 different libraries in play at the same time).

From observation, the only dynamic information about a request that gets passed to the routing layer in Ribbon is the _RibbonApacheHttpRequest_. The _RequestContext_, being a ThreadLocal, is not usable beyond the Hystrix barrier in Zuul's default _SEMAPHORE_ isolation model. This doesn't seem to be the right place to add tags for dynamic routing.

So, that's the first challenge, how would you recommend getting the dynamic request information into the _LoadBalancer_ so that it can make routing decisions based on it?

I am thinking that we could leverage the recently added ServerGroup on the ConsulServer to organize the services by version.

Once we have the Server's tagged with their version and the requested version available to the routing process, the next question is how to actually connect the two. My initial thoughts are to implement this as separate _ServerLists_, as is currently being done by the _ZoneAwareLoadBalancer_ for zones. Ideally, we could then combine the two, so we could route based on both group and zone. This sounds like it could get very tricky though.

One area that is particularly tricky from all the indirection is how a given request gets mapped to a given ServerList for that server. Any clarification on how this work be very helpful.

Is this along the lines you were thinking or did you have a different approach in mind?

I appreciate you taking the time to read my message. I would really appreciate any help that you can provide.

Thanks very much,

Daryl Robbins

**Comment by spencergibb** [comment]
Have you seen this conversation? https://github.com/spring-cloud/spring-cloud-netflix/issues/1290

"not supported" in this case doesn't mean that you can't do it yourself. 

There are many ways to accomplish this.

One way is both versions would go through the same route. The percentage of traffic going to version NEXT would be based on the number of instances: 8 of CURR, 2 of NEXT = 20%. If testing goes as planned, increase NEXT to 8, then put all CURR instances in [maintenance mode](/v1/agent/service/maintenance/). The only requirement for zuul is retry which works if using the deprecated netflix `RestClient`. This is the basic method described by Netflix engineers at SpringOne https://www.infoq.com/presentations/spring-cloud-netflix

**Comment by spencergibb** [comment]
Also see https://jmnarloch.wordpress.com/2015/11/25/spring-cloud-ribbon-dynamic-routing/

**Comment by darylrobbins** [comment]
Thank you, sir. Greatly appreciate the insight. I saw about 4-5 issues on the spring-cloud-consul project, but had missed the one on the spring-cloud-netflix project.

**Comment by darylrobbins** [comment]
Does Pivotal offer any Professional Services offerings for developing this kind of new functionality for us? We are interested in exploring these options to compare against the cost of developing this functionality ourselves? Thanks!
